pipelines/controller/src/main/java/com/google/fhir/analytics/FlinkConfiguration.java

```diff
@@ -187,6 +189,51 @@ private void deleteFilesRecursivelyOnExit(Path path) {
                 }));
   }
 
+  /**
+   * Validates if the JVM memory is sufficient to run pipelines successfully. This validation helps
+   * the application to fail-fast rather than failing during batch execution.
+   */
+  private void validateHeapMemoryConfig(DataProperties dataProperties) {
+    long maxMemory = Runtime.getRuntime().maxMemory();
+    long minJVMMemoryRequired = minJVMMemory(dataProperties);
+    if (minJVMMemoryRequired > maxMemory) {
+      throw new IllegalConfigurationException(
+          String.format(
+              "Insufficient max JVM memory, required %s mb but provided %s mb",
+              memoryInMB(minJVMMemoryRequired), memoryInMB(maxMemory)));
+    }
+  }
+
+  /**
+   * The minimum JVM memory needed for the pipelines to run without fail. This is derived based on
+   * the number of parquet row groups that can be read or written in parallel by the pipeline
+   * threads. This is roughly derived by the below formula.
+   *
+   * <p>Minimum JVM Memory = Memory for JVM stack, perm files etc + (#Parallel Pipeline Threads *
+   * #Parallel Pipelines * (Parquet Row Group Size + buffer for parsing the parquet file))
+   */
+  private long minJVMMemory(DataProperties dataProperties) {
+    int rowGroupSize =
+        dataProperties.getRowGroupSizeForParquetFiles() > 0
+            ? dataProperties.getRowGroupSizeForParquetFiles()
+            : ParquetWriter.DEFAULT_BLOCK_SIZE;
+    int parallelism =
+        dataProperties.getNumThreads() > 0 ? dataProperties.getNumThreads() : DEFAULT_PARALLELISM;
```

Comment:
Does this depend on `dataProperties.getMaxWorkers()` too? BTW, this reminds of [this question](https://github.com/google/fhir-data-pipes/issues/777#issuecomment-1806142211), do you mind replying that too (thanks)?

Reply:
Thanks for the reminder on [this question](https://github.com/google/fhir-data-pipes/issues/777#issuecomment-1806142211), as per the investigation we set the flink `maxParallelism` based on the `dataProperties.getMaxWorkers()` value. And the `maxParallelism` value controls the number of keyed states or the number of partitions in a unbounded data source pipelines i.e a streaming environment. So, we need see if the `dataProperties.getMaxWorkers()` is adding any value here. I have created this [issue](https://github.com/google/fhir-data-pipes/issues/948) to analyze more on this